File: challenges/223/python/solution.py
#!/usr/bin/env python
from challenges.utils import expect


class Solution:
    def computeArea(
        self, ax1: int, ay1: int, ax2: int, ay2: int, bx1: int, by1: int, bx2: int, by2: int
    ) -> int:
        area_of_a = (ay2 - ay1) * (ax2 - ax1)
        area_of_b = (by2 - by1) * (bx2 - bx1)

        # calculate x overlap
        left = max(ax1, bx1)
        right = min(ax2, bx2)
        x_overlap = right - left

        # calculate y overlap
        top = min(ay2, by2)
        bottom = max(ay1, by1)
        y_overlap = top - bottom

        overlap_area = 0
        if x_overlap > 0 and y_overlap > 0:
            overlap_area = x_overlap * y_overlap

        return area_of_a + area_of_b - overlap_area


if __name__ == "__main__":
    obj = Solution()
    expect(obj.computeArea(ax1=-3, ay1=0, ax2=3, ay2=4, bx1=0, by1=-1, bx2=9, by2=2), 45)

File: challenges/223/js/solution.ts
import { expect } from "../../util";

const computeArea = (
  ax1: number,
  ay1: number,
  ax2: number,
  ay2: number,
  bx1: number,
  by1: number,
  bx2: number,
  by2: number
): number => {
  const areaA = (ax2 - ax1) * (ay2 - ay1);
  const areaB = (bx2 - bx1) * (by2 - by1);

  const xOverlap = Math.min(ax2, bx2) - Math.max(ax1, bx1);
  const yOverlap = Math.min(ay2, by2) - Math.max(ay1, by1);
  let overlapArea = 0;
  if (xOverlap > 0 && yOverlap > 0) {
    overlapArea = xOverlap * yOverlap;
  }

  return areaA + areaB - overlapArea;
};

expect(computeArea(-3, 0, 3, 4, 0, -1, 9, 2), 45);
